pandas/core/arrays/categorical.py

```diff
@@ -644,7 +644,11 @@ def from_codes(cls, codes, categories=None, ordered=None, dtype=None):
             )
             raise ValueError(msg)
 
-        codes = np.asarray(codes)  # #21767
+        if is_extension_array_dtype(codes) and is_integer_dtype(codes):
```

Comment:
Hmm does this universally work if we just call `to_numpy`? I suppose could copy where undesired but maybe better than if...else branching

Reply:
Yeah the if / else isn't too pretty but I think we have to worry if `codes` is a numpy array already or possibly a bare list. Ideally it'd be nice to just do `pd.array(codes)` instead of `np.asarray(codes)`, but it looks like `min` and `max` haven't been implemented yet for IntegerArray so we get an error soon after.

Reply:
In principle we could do a `pd.array(codes).to_numpy("int64")`
But a problem with that is that it will convert an already integer array of potentially lower bitsize than int64

Reply:
Has numpy a concept of "any integer dtype"? 
Because then we could also simply do `codes = np.asarray(codes, dtype="any int")`

Another option could be this (to avoid converting an already int ndarray to int64)::

```
if not isinstance(codes, np.ndarray):
    codes = np.asarray(codes, dtype="int64")
```

which seems simpler as what you have now

Reply:
> Another option could be this (to avoid converting an already int ndarray to int64)::
> 
> ```
> if not isinstance(codes, np.ndarray):
>     codes = np.asarray(codes, dtype="int64")
> ```
> 
> which seems simpler as what you have now

I think we want it to raise if codes is something like `[0.0, 1.0, 2.0]` but this would make it a valid input

Reply:
> I think we want it to raise if codes is something like [0.0, 1.0, 2.0] but this would make it a valid input

Ah, yes indeed...

(we should probably try to gather some of those situations we run into handling array-like input including nullable dtypes, to see if we can extract some useful helpers for those cases)